### ISSUE TITLE
Reorganize modules

### ISSUE THREAD

**Issue by yurishkuro** [opentracing/opentracing-java#44]
I propose the following layout
- `opentracing-api` module remains as is
- `opentracing-mock` module remains as is
- `opentracing-impl` is renamed to `opentracing-noop`, and `propagation` package removed from it, or moved to `-impl-java8` as that's the only thing that uses it
- `-impl-java8` is removed from this repo completely and migrated to `basictracer-java` repo, where it can give rise to the actual reference implementation, not just an SPI

**Comment by michaelsembwever** [comment]
-1

Anything that was Java7 compatible is expected to flow from -impl-java8 back to -impl. Originally there was more, but it ended up being just the noop classes. The original agreement around these two was we'll let code be Java8 and if need be (ie an android client came along) we will rewrite the stuff in -impl-java8 back to -impl. Because there was some uncertainty this would ever eventuate it's nicer with -impl-java8.

I'm not sure what you mean as a reference implementation without it being coded against a specific tracing backend. Such code wouldn't exist here anyway, so it makes sense that a supporting APIs around the original API is provided here.

btw opentracing-api is the SPI. The bridge becomes an API. the -impl modules here are supporting APIs to the SPI, or helping to reduce the SPI surface for bridges to implement.

**Comment by yurishkuro** [comment]
> I'm not sure what you mean as a reference implementation without it being coded against a specific tracing backend.

What I mean are these:
  * https://github.com/opentracing/basictracer-go
  * https://github.com/opentracing/basictracer-python
  * https://github.com/opentracing/basictracer-javascript

They are fully functional implementations of OT API with Dapper/Zipkin-style trace context. They are not plugged into any specific backend because nobody needs them to be, but it can be easily done by implementing their internal Recorder interface.

When someone needs to instrument a service or framework for tracing, they need three things only, the `-api.jar`, the `-noop.jar` for default tracer, and the `-mock.jar` for testing their instrumentation. Those are the only 3 things the end users **ever** need to depend on from `opentracing-java`. 

So if the end users do not need the `-impl` modules, and there's not a lot of demand from existing production implementations of the API on these modules, why do we need to keep them in this repo and keep maintaining them? Not to mention that the whole `impl` thing is misleading as they don't actually implement anything, all classes are abstract.

**Comment by michaelsembwever** [comment]
> They are not plugged into any specific backend because nobody needs them to be, but it can be easily done by implementing their internal Recorder interface.

This is an implementation style that is enforced by the OT specification.
While I like it, and I can see benefits to it, it seems just one possible way of doing things.

> Those are the only 3 things the end users ever need to depend on from opentracing-java. 

That's only re-enforcing your premise. If you trying to work off principles then it would be more accurate to say that "The only thing the end users ever need is to depend on opentracing-api." Since it is only the  opentracing-api that can be seen as a requirement to achieve compatibility between  libraries and frameworks that get pulled into individual end user services and applications. Take the libraries and frameworks out of the picture and you don't even need to say opentracing-api is required, because you have the specification.

> So if the end users do not need …

So the end users don't need any of the code, but it's there to help.

> So if the end users do not need the -impl modules, and there's not a lot of demand from existing production implementations of the API on these modules, why do we need to keep them in this repo and keep maintaining them?

That's rather a presumptious thing to say.
I can also state that no implementation has been written without it since it was written.
This type of discussion doesn't get us anything with so few implementations and so little time passed.

> Not to mention that the whole impl thing is misleading as they don't actually implement anything, all classes are abstract.

Not all classes are abstract.
The question of naming is still interesting, but hardly relevant to this issue.

No matter what you call it, it is there to provide guidance and help to implementations, while not being required to use. A basictracer-java implementation can be written and it too can take advantage of it. And they might not, they might have good reason not to or simply suffer NIH.

Having it in the repo though is advantageous as it grounds PRs that are primarily concerned with specifications and higher-level concerns back to java code and what's appropriate within javaland. https://github.com/opentracing/opentracing-java/pull/43 has been a good example of that.

**Comment by yurishkuro** [comment]
The opinions I offer are based on:
  1. having implemented OT tracers in four languages and having them used by over 500 services in production
  1. having instrumented a couple dozen of internal and open source frameworks with OT API

So I have a pretty well rounded experience to know what the API consumers need.
  1. On instrumentation side (end users), the Noop tracer is important as the default configuration, unless they want to code around `if (tracer == null)`. The Mock tracer is important for unit tests (NB: it’s not the same mock as provided by mocking frameworks, it’s a fully functional tracer). Testing can also use real tracer implementations like Jaeger or Lightstep, but those tend to have much heavier dependencies than the Mock tracer.
  1. On implementation side, the reference implementations (basictracer-xxx) have been quite useful, both as playgrounds for POCs and for collaborative performance experiments/optimizations. The abstract/skeleton implementations have not been useful. We actually used to have those in the other languages early on, but quickly realized that (a) there aren’t that many people implementing tracers to benefit from shared implementation, and (b) implementers tend to have strong opinions about performance that affects how code is structured very close to the API, so a common layer with hardcoded performance implications was not welcomed.

I don’t object to keeping the abstract skeletons, but I think we need to:
  a. rename the modules to reflect what they are, as they are not impls and it is confusing to people visiting this repo
  b. have an actual tracer based on those skeletons (a Mock tracer would be a good candidate)

In summary, I don’t get a feeling we’re going to get traction on this issue directly, so I am splitting it into smaller pieces, please continue comments on those.
 * move Noop tracer to opentracing-noop module #45
 * rename `-impl` and `-impl-java8` to reflect their true content  #46
 * refactor Mock tracer to be based on former `-impl`

**Comment by michaelsembwever** [comment]
As I've stated in each individual issue I think this is all just a storm in a teacup.

`-impl.jar` alongside a `-api.jar` is a common enough precedent in java without it implying a finished implementation or bridge. OT is a SPI so it's a given that you need a bridge to complete it.

`-impl` currently only contains the noop classes. Neither `impl` nor `-impl-java8` has transitive dependencies. So again why are we spending effort on this?

The Mock tracer can use the `-impl-java8` classes. The Brave bridge uses it.

I look forward to seeing the basictracer-java code, but don't see the point in what is imo premature optimisation.

**Comment by bensigelman** [comment]
I was reminded of this issue today when @mbroad was asking some questions about the -impl artifact.

I would prefer that the `github.com/opentracing/opentracing-java` repo only contained code that OpenTracing *callers* reference directly. This is the way things work in every other language repo and I don't see why things should be different in Java, esp since the API has stabilized. Regardless of what we actually call the various other artifacts, they are a better fit for the `github.com/opentracing-contrib` charter: potentially useful code (whether for Tracer implementations or for OSS integrations or whatever else) that doesn't change the core OpenTracing API.

**Comment by bhs** [comment]
This is either done or infeasible now (depending on how one looks at it).
